core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1597,6 +1640,327 @@ public void addStateChangeListener(StateChangeListener<DistributedStagesSchedule
+                DistributedStagesSchedulerStateMachine stateMachine,
+                QueryStateMachine queryStateMachine,
+                List<FaultTolerantStageScheduler> schedulers,
+                SplitSchedulerStats schedulerStats,
+                NodeAllocator nodeAllocator,
+                ScheduledFuture<?> nodeUpdateTask)
+        {
+            this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
+            this.queryStateMachine = requireNonNull(queryStateMachine, "queryStateMachine is null");
+            this.schedulers = requireNonNull(schedulers, "schedulers is null");
+            this.schedulerStats = requireNonNull(schedulerStats, "schedulerStats is null");
+            this.nodeAllocator = requireNonNull(nodeAllocator, "nodeAllocator is null");
+            this.nodeUpdateTask = requireNonNull(nodeUpdateTask, "nodeUpdateTask is null");
+        }
+
+        @Override
+        public void schedule()
+        {
+            checkState(started.compareAndSet(false, true), "already started");
+
+            try (SetThreadName ignored = new SetThreadName("Query-%s", queryStateMachine.getQueryId())) {
+                List<ListenableFuture<Void>> blockedStages = new ArrayList<>();
+                while (!isFinishingOrDone(queryStateMachine) && !stateMachine.getState().isDone()) {
+                    blockedStages.clear();
+                    boolean atLeastOneStageIsNotBlocked = false;
+                    boolean allFinished = true;
+                    for (FaultTolerantStageScheduler scheduler : schedulers) {
+                        if (scheduler.isFinished()) {
+                            continue;
+                        }
+                        allFinished = false;
+                        ListenableFuture<Void> blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                            continue;
+                        }
+                        try {
+                            scheduler.schedule();
+                        }
+                        catch (Throwable t) {
+                            stateMachine.transitionToFailed(t, Optional.of(scheduler.getStageId()));
+                            return;
+                        }
+                        blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                        }
+                        else {
+                            atLeastOneStageIsNotBlocked = true;
+                        }
+                    }
+                    if (allFinished) {
+                        stateMachine.transitionToFinished();
+                        return;
+                    }
+                    // wait for a state change and then schedule again
+                    if (!atLeastOneStageIsNotBlocked && !blockedStages.isEmpty()) {
```

Comment:
Is that possbile to have `!atLeastOneStageIsNotBlocked` and empty empty `blockedStages`? It looks that would mean that we have no schedulers at all. 
This should probably be rejected in the contractor.
Then we should change the `&& !blockedStages.isEmpty()` to `verify(!blockedStages.isEmpty())`

Reply:
Is it possible to have a query with no distributed stages (when all the stages are "coordinator only")? In this scenario there could be no stages to schedule. Do you think it makes sense to special case it and not create a distributed scheduler at all? I feels like it may unnecessarily increase complexity.

Reply:
Added a special case in the `schedule` method to avoid busy looping.